File: light-control/command_watcher.py
import json
import logging
from typing import List
import zmq

class CommandWatcher:
    def __init__(self, available_commands: List[str], queue_port=5555):
        self.commands_received = []
        self._available_commands = available_commands
        self.__logger = logging.getLogger()
        self.__initialize_queue(queue_port)
    
    def destroy(self):
        self.__poller.unregister(self.__socket)
        self.__socket.close()
        self.__context.destroy()
    
    def __initialize_queue(self, port):
        self.__poller = zmq.Poller()
        self.__context = zmq.Context()
        self.__socket = self.__context.socket(zmq.REP)
        self.__socket.bind('tcp://*:{0}'.format(port))
        self.__poller.register(self.__socket, zmq.POLLIN)

        self.__logger.log(logging.INFO, 'socket bound')

    def check_messages(self):
        results = self.__poller.poll(0)

        for socket, event_mask in results:
            if event_mask == zmq.POLLIN:
                message = socket.recv_json(zmq.DONTWAIT)
                self.handle(message)

    def handle(self, message):
        self.__logger.log(logging.DEBUG, 'Message received %s', message)
        
        command = message['command'] if 'command' in message else ''
        if command in self._available_commands:
            if command not in self.commands_received:
                self.commands_received.append(command)
            self.__socket.send_json({
                'accepted': True,
                'message': 'Good Job',
            })
        else:
            self.__socket.send_json({
                'accepted': False,
                'message': 'Command not recognized',
            })

    def mark_complete(self, command:str):
        self.commands_received.remove(command)

File: light-control-ts/src/commandWatcher.ts
import * as zmq from "zeromq";

export interface Command {
    command: string;
    [key: string]: unknown;
}

const AVAILABLE_COMMANDS = ["next", "off", "on", "set_brightness", "set_cycle", "get_status", "get_cycles"] as const;

export class CommandWatcher {
    commandsReceived: Command[] = [];
    private readonly socket: zmq.Reply;
    // Track the pending reply function for the current in-flight request
    private pendingReply: ((response: unknown) => Promise<void>) | null = null;

    constructor(port = 5555) {
        this.socket = new zmq.Reply();
        this.socket.bind(`tcp://*:${port}`).then(() => {
            console.log(`ZMQ socket bound on port ${port}`);
            this.receiveLoop();
        });
    }

    private async receiveLoop(): Promise<void> {
        for await (const [msgBuffer] of this.socket) {
            let message: Command;
            try {
                message = JSON.parse(msgBuffer.toString()) as Command;
            } catch {
                await this.socket.send(JSON.stringify({ accepted: false, message: "Invalid JSON" }));
                continue;
            }

            const command = message.command ?? "";
            if (!(AVAILABLE_COMMANDS as readonly string[]).includes(command)) {
                await this.socket.send(JSON.stringify({ accepted: false, message: "Command not recognized" }));
                continue;
            }

            // Park the reply capability and enqueue for main loop processing
            let resolveReply!: (response: unknown) => void;
            const replyPromise = new Promise<unknown>((resolve) => {
                resolveReply = resolve;
            });

            this.pendingReply = async (response: unknown) => {
                resolveReply(response);
            };
            this.commandsReceived.push(message);

            // Wait for the main loop to call markComplete before accepting the next message
            const response = await replyPromise;
            await this.socket.send(JSON.stringify({ accepted: true, response }));
            this.pendingReply = null;
        }
    }

    checkMessages(): void {
        // Queue is already populated by receiveLoop — nothing to do here
    }

    async markComplete(message: Command, response: unknown = null): Promise<void> {
        const idx = this.commandsReceived.indexOf(message);
        if (idx !== -1) this.commandsReceived.splice(idx, 1);
        if (this.pendingReply) await this.pendingReply(response);
    }

    destroy(): void {
        this.socket.close();
    }
}
